src/modm/platform/can/stm32-fdcan/message_ram.hpp

```diff
@@ -0,0 +1,270 @@
+		return reinterpret_cast<uint32_t*>(address);
+	}
+public:
+	/// Write TX element headers to TX queue
+	static void
+	writeTxHeaders(uint8_t putIndex, CommonFifoHeader_t common, TxFifoHeader_t tx)
+	{
+		uint32_t* messageRam = txFifoElement(putIndex);
+		*messageRam++ = common.value;
+		*messageRam = tx.value;
+	}
+
+	/// Read RX element headers from RX fifo
+	static std::tuple<CommonFifoHeader, RxFifoHeader>
+	readRxHeaders(RxFifoAddress address)
+	{
+		const uint32_t* messageRam = address.ptr();
+		const auto commonHeader = CommonFifoHeader{*messageRam++};
+		const auto rxHeader = RxFifoHeader{*messageRam};
+		return {commonHeader, rxHeader};
+	}
+
+	/// Read message data from RX fifo
+	static void
+	readData(RxFifoAddress address, std::span<uint8_t> outData)
+	{
+		const auto size = std::min(outData.size(), 64u);
+
+		// + 2: skip 2x32 bit headers
+		const uint32_t* messageRam = address.ptr() + 2;
+		// message data must be read in 32bit accesses, memcpy on message ram does not work
+		for (auto i = 0u; i < size; i += 4) {
+			const uint32_t receivedData = *messageRam++;
+
+			const auto byteCount = std::min(4u, size - i);
+			std::memcpy(&outData[i], &receivedData, byteCount);
+		}
+	}
+
+	/// Write message data to TX queue
+	static void
+	writeData(uint_fast8_t putIndex, std::span<const uint8_t> inData)
+	{
+		const auto size = std::min(inData.size(), 64u);
+
+		// + 2: skip 2x32 bit headers
+		uint32_t* messageRam = txFifoElement(putIndex) + 2;
+		// message data must be written in 32bit accesses, memcpy to message ram does not work
+		for (auto i = 0u; i < size; i += 4) {
+			const auto byteCount = std::min(4u, size - i);
+
+			uint32_t outputData{};
+			std::memcpy(&outputData, &inData[i], byteCount);
+
+			*messageRam++ = outputData;
+		}
+	}
+
+	/// Construct common fifo header from CanMessage
+	static CommonFifoHeader_t
+	headerFromMessage(const modm::can::Message& message)
+	{
+		CommonFifoHeader_t header = (message.isExtended() ? CommonFifoHeader::ExtendedId : CommonFifoHeader(0))
+			| (message.isRemoteTransmitRequest() ? CommonFifoHeader::RemoteFrame : CommonFifoHeader(0));
+
+		const auto canId = message.isExtended() ? message.getIdentifier() : (message.getIdentifier() << 18);
+		CanId_t::set(header, canId);
+
+		return header;
+	}
+
+	static void
+	setStandardFilter(uint8_t index, FilterType type, FilterConfig config, uint16_t id1, uint16_t id2)
+	{
+		constexpr auto idMask = (1u << 11) - 1;
+		*standardFilter(index) = uint32_t(type) | uint32_t(config) |
+			(id2 & idMask) | ((id1 & idMask) << 16);
+	}
+
+	static void
+	setExtendedFilter0(uint8_t index, FilterConfig config, uint32_t id)
+	{
+		constexpr auto idMask = (1u << 29) - 1;
+		*extendedFilter(index) = (uint32_t(config) << 2) | (id & idMask);
+	}
+
+	static void
+	setExtendedFilter1(uint8_t index, FilterType type, uint32_t id)
+	{
+		constexpr auto idMask = (1u << 29) - 1;
+		*(extendedFilter(index) + 1) = uint32_t(type) | (id & idMask);
```

Comment:
Unaligned access is not an issue with the message ram?!

Reply:
This is not unaligned access. This is incrementing a `uint32_t` by one, thus the address is incremented by 4.